src/utils/contracts/AdapterInvestLimitBase.sol

```diff
@@ -0,0 +1,113 @@
+
+    /** @notice max deposit value datatypes */
+    MaxExposure public maxDepositProtocolMode;
+
+    /** @notice max deposit's protocol value in percentage */
+    uint256 public maxDepositProtocolPct; // basis points
+
+    /** @notice  Maps liquidityPool to max deposit value in percentage */
+    mapping(address => uint256) public maxDepositPoolPct; // basis points
+
+    /** @notice  Maps liquidityPool to max deposit value in absolute value for a specific token */
+    mapping(address => mapping(address => uint256)) public maxDepositAmount;
+
+    constructor() public {
+        maxDepositProtocolPct = uint256(10000); // 100% (basis points)
+        maxDepositProtocolMode = MaxExposure.Pct;
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositPoolPct(address _liquidityPool, uint256 _maxDepositPoolPct)
+        external
+        override
+        onlyRiskOperator
+    {
+        maxDepositPoolPct[_liquidityPool] = _maxDepositPoolPct;
+        emit LogMaxDepositPoolPct(maxDepositPoolPct[_liquidityPool], msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositAmount(
+        address _liquidityPool,
+        address _underlyingToken,
+        uint256 _maxDepositAmount
+    ) external override onlyRiskOperator {
+        maxDepositAmount[_liquidityPool][_underlyingToken] = _maxDepositAmount;
+        emit LogMaxDepositAmount(maxDepositAmount[_liquidityPool][_underlyingToken], msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositProtocolMode(MaxExposure _mode) external override onlyRiskOperator {
+        maxDepositProtocolMode = _mode;
+        emit LogMaxDepositProtocolMode(maxDepositProtocolMode, msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositProtocolPct(uint256 _maxDepositProtocolPct) external override onlyRiskOperator {
+        maxDepositProtocolPct = _maxDepositProtocolPct;
+        emit LogMaxDepositProtocolPct(maxDepositProtocolPct, msg.sender);
+    }
+
+    /**
+     * @dev Get the final value of amount in underlying token to be deposited
+     * @param _liquidityPool liquidity pool address
+     * @param _underlyingToken underlying token address
+     * @param _amount amount in underlying token
+     * @param _poolValue pool value
+     * @return amount in underlying token to be deposited affected by investment limitation
+     */
+    function _getDepositAmount(
+        address _liquidityPool,
+        address _underlyingToken,
+        uint256 _amount,
+        uint256 _poolValue
+    ) internal view returns (uint256) {
+        uint256 _limit =
+            maxDepositProtocolMode == MaxExposure.Pct
+                ? _getMaxDepositAmountByPct(_liquidityPool, _poolValue)
+                : maxDepositAmount[_liquidityPool][_underlyingToken];
+        return _amount > _limit ? _limit : _amount;
+    }
```

Comment:
- [ ] If `_getDepositAmount` function is made public and extended as a part of the `IAdapterInvestLimitV2` interface Vault can benefit our of it by making sure deposit and withdraw actually happened.

(Note : - There are protocols like Compound and forks of Compound which return some error value on failure while doing deposit/withdraw. In this scenario, pre calculating the amount that will be deposited to protocol will ensure robustness of `rebalance` and `vaultDepositAllToStrategy` functions)

```
uint256 _limit =
            maxDepositProtocolMode == MaxExposure.Pct
                ? _getMaxDepositAmountByPct(_liquidityPool, _poolValue)
                : maxDepositAmount[_liquidityPool][_underlyingToken];
```
In the above formulae, allAmountInToken should be subtracted `maxDepositAmount[_liquidityPool][_underlyingToken]` considering previous vault deposits if any.